### PR TITLE
OC-391 Adding new versions

### PR DESCRIPTION
**Pull request by florin-holhos**
The purpose of this PR was to implement adding new versions and updating their DOIs when publishing a new version. 
Stage checks have been added so that the new functionality si currently available just on _int_ and _local_ 

- added new "draft" column for the "Links" model in prisma
- moved topics from "Publication" to "PublicationVersion" model and created migration script to run before deploying the latest DB schema.

**Important!**
Deploying the latest DB schema directly will erase existing topics from existing Publications.

Deploy steps:

1. Put the website into maintenance mode from AWS Amplify
2. Connect to the corresponding int/prod EC2 instance and cd into octopus project
3. checkout to the corresponding branch int/prod and "git pull" to bring the latest changes
4. Checkout to first commit: a4b3b47605146df06af37b62631d0c174c329fc1
5. Apply DB migration which will add "draft" column on the "Link" model in prisma
6. Run "updateLinksStatus" script in order to update "draft" column for each link
7. Check that the links "draft" column is updated correctly: _true_ for DRAFTs and _false_ for LIVE
8. Checkout to second commit: 207f113a88ec6d4b6759bf5558c586682d12867a
9. Apply DB migration which will add 'topics' column on the "PublicationVersion" model 
10. Run "migrateTopicsFromPubsToVersions" script which will add "topics" from publications to each corresponding v1
11. Check that topics were correctly copied from publications to publication versions
12. Checkout back to the branch itself int/prod
13. Apply the last DB migration which will remove "topics" column from "Publication" model
14. Done!

_Note!_ - need to run "updateLinksStatus" before running e2e tests, some tests are checking linked publications and those publications will not be visible in the UI if their "draft" status was not updated accordingly.

---

### Acceptance Criteria:

As per [OC-391](https://jiscdev.atlassian.net/browse/OC-391)

---

### Checklist:

- [x] Local manual testing conducted
- [x] Automated tests added

---

### Tests:

---

### Screenshots:

[OC-391]: https://jiscdev.atlassian.net/browse/OC-391?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: e2e/tests/LoggedIn/publish.e2e.spec.ts [hunk 1]
```diff
@@ -446,14 +443,13 @@ test.describe('Publication flow', () => {
 
         // Preview and check preview draft publication
         await page.locator(PageModel.publish.previewButton).click();
-        await checkPublication(page, hypothesisPublication);
+        await checkPublication(page, hypothesisPublication, [Helpers.user1]);
 
         // Publish and check live publication
         await page.locator(PageModel.publish.draftEditButton).click();
-        await page.waitForResponse((response) => response.url().includes('/references') && response.ok());
```
Comment (finlay-jisc) [inline e2e/tests/LoggedIn/publish.e2e.spec.ts:453]: These "waitForResponse" calls seemed to be keeping the tests more stable - from trying the tests out on this branch they are being flakier than before. Can we keep them?

File: api/src/components/publicationVersion/__tests__/createPublicationVersion.test.ts [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+            .query({
+                apiKey: '123456789'
+            });
+
+        expect(newPublicationVersion.status).toEqual(201);
+        expect(newPublicationVersion.body.currentStatus).toEqual('DRAFT');
+        expect(newPublicationVersion.body.versionNumber).toEqual(2);
+        expect(newPublicationVersion.body.title).toEqual(oldPublicationVersion.body.title);
+        expect(newPublicationVersion.body.content).toEqual(oldPublicationVersion.body.content);
+        expect(newPublicationVersion.body.topics.length).toEqual(oldPublicationVersion.body.topics.length);
```
[the remark this answers is not in the document]
Comment (florin-holhos) [reply api/src/components/publicationVersion/__tests__/createPublicationVersion.test.ts:44]: Added a new commit for this

File: api/src/components/publicationVersion/__tests__/createPublicationVersion.test.ts [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+    });
+
+    test('User cannot create a new publication version if they are not an author on the latest version.', async () => {
+        const newPublicationVersion = await testUtils.agent
+            .post('/publications/publication-data-draft/publication-versions')
+            .query({
+                apiKey: '000000005'
+            });
+
+        expect(newPublicationVersion.status).toEqual(403);
```
[the remark this answers is not in the document]
Comment (florin-holhos) [reply api/src/components/publicationVersion/__tests__/createPublicationVersion.test.ts:68]: The publication id was wrong. Added a new commit for that.

File: api/src/components/publicationVersion/__tests__/updateStatus.test.ts [hunk 4]
```diff
@@ -254,4 +254,41 @@ describe('Update publication version status', () => {
...
+        // create a new version
+        const newPublicationVersion = await testUtils.agent
+            .post('/publications/publication-problem-live-2/publication-versions')
+            .query({
+                apiKey: '123456789'
+            });
+
+        expect(newPublicationVersion.status).toEqual(201);
+        expect(newPublicationVersion.body.currentStatus).toEqual('DRAFT');
+        expect(newPublicationVersion.body.doi).toEqual(null); // DOI is only generated when the new version goes LIVE
```
[the remark this answers is not in the document]
Comment (florin-holhos) [reply api/src/components/publicationVersion/__tests__/updateStatus.test.ts]: We don't really need to create a new version via de API but this was a quick way (and less code) to test the entire flow and make sure everything works fine. I've mocked a v2 for this publication and just update it's status to 'LIVE', although, arguably the previous code was testing the entire flow and was making sure that creation + publishing a new version works fine from start to finish.

File: api/src/components/publicationVersion/__tests__/updateTopics.test.ts [hunk 5]
```diff
@@ -34,44 +34,44 @@ describe('Update publication topics', () => {
...
             .query({ apiKey: '987654321' })
             .send({ topics: ['test-topic-1'] });
 
         expect(updateTopicsRequest.status).toEqual(403);
         expect(updateTopicsRequest.body.message).toEqual(
-            'You do not have permission to update topics for this publication.'
+            'You do not have permission to update topics for this publication version.'
         );
     });
 });
```
[the remark this answers is not in the document]
Comment (florin-holhos) [reply api/src/components/publicationVersion/__tests__/updateTopics.test.ts:77]: Took me a while to understand what was the concern with this :)) 
Moved update topics tests into updatePublicationVersion.test.ts file.

File: api/src/components/publicationVersion/controller.ts [hunk 6]
```diff
@@ -85,19 +110,19 @@ export const update = async (
...
-            return response.json(403, {
+            return response.json(404, {
                 message: 'This publication version does not exist.'
             });
         }
 
         if (publicationVersion.user.id !== event.user.id) {
             return response.json(403, {
-                message: 'You do not have permission to modify this publication version.'
+                message: 'You do not have permission to update topics for this publication version.'
```
[the remark this answers is not in the document]
Comment (florin-holhos) [reply api/src/components/publicationVersion/controller.ts:120]: Hmm, I don't remember making this change. I think it came from rebase and fixing conflicts. Fixed now anyway.

File: api/src/components/publicationVersion/controller.ts [hunk 7]
```diff
@@ -26,18 +26,43 @@ export const get = async (
             return response.json(200, publicationVersion);
         }
 
-        // only the owner or co-authors can view publications
+        // only the owner or co-authors can view a DRAFT publication version
         if (
-            event.user?.id === publicationVersion.user.id ||
-            publicationVersion.coAuthors.some((coAuthor) => coAuthor.linkedUser === event.user?.id)
+            !(
```
[the remark this answers is not in the document]
Comment (florin-holhos) [reply api/src/components/publicationVersion/controller.ts:31]: This check is working as expected from what I see.
Unconfirmed co-author is getting a 404
image
image

After co-author confirms their involvement they need to be able to see the pub version
image
image

While the draft is being edited by the corresponding author
image

File: api/src/components/publicationVersion/__tests__/updateStatus.test.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (finlay-jisc) [reply api/src/components/publicationVersion/__tests__/updateStatus.test.ts]: OK, no worries

File: api/src/components/publicationVersion/controller.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (finlay-jisc) [reply api/src/components/publicationVersion/controller.ts:31]: The data can still get through from the API though. For example in the last scenario:
Screenshot 2023-11-15 163208

If the corresponding author has put the publication back into draft status, from locked, coauthors can still see the draft data even though it isn't ready for them to approve yet. We don't allow this before a publication is locked so I think we shouldn't allow it afterwards either, if the current status is draft

[the remark this answers is not in the document]
Comment (florin-holhos) [reply api/src/components/publicationVersion/controller.ts:31]: But this is how it always worked. Once they confirmed their involvement they could see the DRAFT/LOCKED publication. Nothing changed about how this works.

File: ui/src/pages/publications/[id]/edit.tsx [hunk 8]
```diff
@@ -82,34 +82,24 @@ const steps: Types.CreationSteps = {
...
+    const promises: [Promise<Interfaces.PublicationVersion | void>, Promise<Interfaces.PublicationWithLinks | void>] = [
+        api
+            .get(`${Config.endpoints.publications}/${publicationId}/publication-versions/latest`, token)
+            .then((res) => res.data)
+            .catch((error) => console.log(error)),
+
+        api
+            .get(`${Config.endpoints.publications}/${publicationId}/links?direct=true`, token)
+            .then((res) => res.data)
+            .catch((error) => console.log(error))
```
[the remark this answers is not in the document]
Comment (florin-holhos) [reply ui/src/pages/publications/[id]/edit.tsx]: Yep, re-organized fetched data SSR

File: api/src/components/publicationVersion/controller.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (finlay-jisc) [reply api/src/components/publicationVersion/controller.ts:31]: Ok, let's leave it

File: e2e/tests/LoggedIn/publish.e2e.spec.ts (hunk 1 above)
Comment (florin-holhos) [reply e2e/tests/LoggedIn/publish.e2e.spec.ts:453]: The api call for /references has been moved from CSR to SSR. That means there's no http call to /references from the client anymore, which makes all tests to fail waiting for response.
image

Comment (finlay-jisc) [reply e2e/tests/LoggedIn/publish.e2e.spec.ts:453]: Ah I see, that's ok then

**Review by finlay-jisc — APPROVED** [review]

**Pull request merged**